src/kem/kyber/ntt.rs

```diff
@@ -85,80 +122,56 @@ pub(in crate::kem::kyber) fn ntt_binomially_sampled_ring_element(
-    ntt_at_layer!(4, zeta_i, re, 3328);
-    ntt_at_layer!(3, zeta_i, re, 3328);
-    ntt_at_layer!(2, zeta_i, re, 3328);
-    ntt_at_layer!(1, zeta_i, re, 3328);
+    (re, zetas_index) = ntt_at_layer(7, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = ntt_at_layer(6, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = ntt_at_layer(5, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = ntt_at_layer(4, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = ntt_at_layer(3, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = ntt_at_layer(2, zetas_index, re, initial_coefficient_bound);
+    (re, _) = ntt_at_layer(1, zetas_index, re, initial_coefficient_bound);
 
     re.coefficients = re.coefficients.map(barrett_reduce);
 
     re
 }
 
 #[inline(always)]
-fn invert_ntt_montgomery<const K: usize>(
-    mut re: KyberPolynomialRingElement,
-) -> KyberPolynomialRingElement {
+pub(crate) fn invert_ntt_montgomery<const K: usize>(
+    mut re: PolynomialRingElement,
+) -> PolynomialRingElement {
     // We only ever call this function after matrix/vector multiplication
+    let initial_coefficient_bound = (K as i32) * FIELD_MODULUS;
     hax_lib::debug_assert!(re
         .coefficients
         .into_iter()
-        .all(|coefficient| coefficient.abs() < (K as i32) * 3329));
-
-    let mut zeta_i = COEFFICIENTS_IN_RING_ELEMENT / 2;
-
-    macro_rules! invert_ntt_at_layer {
-        ($layer:literal) => {
-            let step = 1 << $layer;
+        .all(|coefficient| coefficient.abs() < initial_coefficient_bound));
 
-            for round in 0..(128 / step) {
-                zeta_i -= 1;
+    let mut zetas_index = COEFFICIENTS_IN_RING_ELEMENT / 2;
 
-                let offset = round * step * 2;
-
-                for j in offset..offset + step {
-                    let a_minus_b = re.coefficients[j + step] - re.coefficients[j];
-
-                    // Instead of dividing by 2 here, we just divide by
-                    // 2^7 in one go in the end.
-                    re.coefficients[j] = re.coefficients[j] + re.coefficients[j + step];
-                    re.coefficients[j + step] =
-                        montgomery_reduce(a_minus_b * ZETAS_MONTGOMERY_DOMAIN[zeta_i]);
-                }
-            }
-        };
-    }
+    (re, zetas_index) = invert_ntt_at_layer(1, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = invert_ntt_at_layer(2, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = invert_ntt_at_layer(3, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = invert_ntt_at_layer(4, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = invert_ntt_at_layer(5, zetas_index, re, initial_coefficient_bound);
+    (re, zetas_index) = invert_ntt_at_layer(6, zetas_index, re, initial_coefficient_bound);
+    (re, _) = invert_ntt_at_layer(7, zetas_index, re, initial_coefficient_bound);
 
```

Comment:
Not using mutability for `re` and `zetas_index` in these functions might make the extaction nicer.
I also see a slight performance increase with that change. But it's pretty noisy so I'm not sure if it's real.

Reply:
I had to revert this change and convert it back to a macro, since hax does not extract this properly.